// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package protocol.control_service;

import "common.proto";

/*
 * Used for low-level control of a vehicle
 */
service Control {
  // Connects to the vehicle
  rpc Connect (ConnectRequest) returns (ConnectResponse) {}
  // Checks whether the vehicle is successfully connected
  rpc IsConnected (IsConnectedRequest) returns (IsConnectedResponse) {}
  // Disconnects from the vehicle
  rpc Disconnect (DisconnectRequest) returns (DisconnectResponse) {}
  // Order the vehicle to arm
  rpc Arm (ArmRequest) returns (ArmResponse) {}
  // Order the vehicle to disarm
  rpc Disarm (DisarmRequest) returns (DisarmResponse) {}
  // Order the vehicle to take off
  rpc TakeOff (TakeOffRequest) returns (stream TakeOffResponse) {}
  // Land the vehicle at its current position
  rpc Land (LandRequest) returns (stream LandResponse) {}
  // Order the vehicle to hold/loiter
  rpc Hold (HoldRequest) returns (stream HoldResponse) {}
  // Emergency shutdown of the vehicle motors
  rpc Kill (KillRequest) returns (stream KillResponse) {}
  // Changes the home destination for the vehicle
  rpc SetHome (SetHomeRequest) returns (SetHomeResponse) {}
  // Return to the vehicle home destination
  rpc ReturnToHome (ReturnToHomeRequest) 
    returns (stream ReturnToHomeResponse) {}
  // Transit the vehicle to a target global position, expressed
  // in global coordinates
  rpc SetGlobalPosition (SetGlobalPositionRequest)
    returns (stream SetGlobalPositionResponse) {}
  // Transit the vehicle to a target position relative to
  // the global ENU (East, North, Up) or vehicle frame of 
  // reference, in meters
  rpc SetRelativePosition (SetRelativePositionRequest)
    returns (stream SetRelativePositionResponse) {}
  // Transit the vehicle at a target velocity in the global
  // ENU (East, North, Up) or vehicle frame of reference, 
  // in meters per second
  rpc SetVelocity (SetVelocityRequest)
    returns (stream SetVelocityResponse) {}
  // Sets the heading of the vehicle
  rpc SetHeading (SetHeadingRequest)
    returns (stream SetHeadingResponse) {}
  // Set the pose of the target gimbal
  rpc SetGimbalPose (SetGimbalPoseRequest)
    returns (stream SetGimbalPoseResponse) {}
  // Set the vehicle video stream parameters
  rpc ConfigureImagingSensorStream (ConfigureImagingSensorStreamRequest)
    returns (ConfigureImagingSensorStreamResponse) {}
  // Set the vehicle telemetry stream parameters
  rpc ConfigureTelemetryStream (ConfigureTelemetryStreamRequest)
    returns (ConfigureTelemetryStreamResponse) {}   
}

message ConnectRequest {
  protocol.common.Request request = 1;
}

message ConnectResponse {
  protocol.common.Response response = 1;
}

message IsConnectedRequest {
  protocol.common.Request request = 1;
}

message IsConnectedResponse {
  protocol.common.Response response = 1;
  bool is_connected = 2;
}

message DisconnectRequest {
  protocol.common.Request request = 1;
}

message DisconnectResponse {
  protocol.common.Response response = 1;
}

message ArmRequest {
  protocol.common.Request request = 1;
}

message ArmResponse {
  protocol.common.Response response = 1;
}

message DisarmRequest {
  protocol.common.Request request = 1;
}

message DisarmResponse {
  protocol.common.Response response = 1;
}

message TakeOffRequest {
  protocol.common.Request request = 1;
  // Take off height in relative altitude [meters]
  float take_off_altitude = 2;
}

message TakeOffResponse {
  protocol.common.Response response = 1;
}

message LandRequest {
  protocol.common.Request request = 1;
}

message LandResponse {
  protocol.common.Response response = 1;
}

message HoldRequest {
  protocol.common.Request request = 1;
}

message HoldResponse {
  protocol.common.Response response = 1;
}

message KillRequest {
  protocol.common.Request request = 1;
}

message KillResponse {
  protocol.common.Response response = 1;
}

message SetHomeRequest {
  protocol.common.Request request = 1;
  protocol.common.Location location = 2; // New home location
}

message SetHomeResponse {
  protocol.common.Response response = 1;
}

message ReturnToHomeRequest {
  protocol.common.Request request = 1;
}

message ReturnToHomeResponse {
  protocol.common.Response response = 1;
}

enum AltitudeMode {
  ABSOLUTE = 0; // Meters above Mean Sea Level
  RELATIVE = 1; // Meters above takeoff location
}

enum HeadingMode {
  TO_TARGET = 0; // Orient towards the target location
  HEADING_START = 1; // Orient towards the given heading
}

message SetGlobalPositionRequest {
  protocol.common.Request request = 1;
  protocol.common.Location location = 2; // Target location
  // Determines whether the drone will consider altitude as meters above MSL
  // (Mean Sea Level), or relative to its takeoff location.
  AltitudeMode altitude_mode = 4;
  // Determines how the drone will orient during transit 
  HeadingMode heading_mode = 5;
  // Maximum velocity during transit, north_vel determines horizontal
  // velocity, up_vel determines vertical velocity, and angular_vel
  // determines angular velocity
  protocol.common.Velocity max_velocity = 6;
}

message SetGlobalPositionResponse {
  protocol.common.Response response = 1;
}

enum ReferenceFrame {
  BODY = 0; // Vehicle reference frame
  ENU = 1;  // Global (East, North, Up) reference frame
}

message SetRelativePositionRequest {
  protocol.common.Request request = 1;
  // Target position
  protocol.common.Position position = 2;
  // Maximum velocity during transit, x_vel determines horizontal
  // velocity, up_vel determines vertical velocity, and angular_vel
  // determines angular velocity
  protocol.common.Velocity max_velocity = 3;
  // Frame of reference
  ReferenceFrame frame = 4;
}

message SetRelativePositionResponse {
  protocol.common.Response response = 1;
}

message SetVelocityRequest {
  protocol.common.Request request = 1;
  // Target velocity
  protocol.common.Velocity velocity = 2;
  // Frame of reference
  ReferenceFrame frame = 3;
}

message SetVelocityResponse {
  protocol.common.Response response = 1;
}

message SetHeadingRequest {
  protocol.common.Request request = 1;
  // Target heading or global location to look at
  protocol.common.Location location = 2;
  // Determines how the drone will orient
  HeadingMode heading_mode = 5;
}

message SetHeadingResponse {
  protocol.common.Response response = 1;
}

enum PoseMode {
  ANGLE = 0; // Absolute angle
  OFFSET = 1; // Offset from current
  VELOCITY = 2; // Rotational velocities
}

message SetGimbalPoseRequest {
  protocol.common.Request request = 1;
  uint32 gimbal_id = 2; // ID of the target gimbal
  protocol.common.Pose pose = 3; // Target pose
  PoseMode mode = 4; // Specifies how to interpret the target pose
}

message SetGimbalPoseResponse {
  protocol.common.Response response = 1;
}

message ImagingSensorConfiguration {
  uint32 id = 1; // Target imaging sensor ID
  bool set_primary = 2; // Set this sensor as the primary stream
  uint32 set_fps = 3; // Target FPS for stream
} 

message ConfigureImagingSensorStreamRequest {
  protocol.common.Request request = 1;
  // List of configurations to be updated
  repeated ImagingSensorConfiguration configurations = 2;
}

message ConfigureImagingSensorStreamResponse {
  protocol.common.Response response = 1;
}

message ConfigureTelemetryStreamRequest {
  protocol.common.Request request = 1;
  // Target frequency of telemetry generation, in Hz
  uint32 frequency = 2;
}

message ConfigureTelemetryStreamResponse {
  protocol.common.Response response = 1;
}
